### PR TITLE
Fix of tensor names loosing for cf=True flag in MOC transformations.

### PR DESCRIPTION
**Pull request by popovaan**
Root cause analysis: 
Names are lost after `validate_nodes_and_infer_types()`. 
In Result `validate_and_infer_types()` output tensor pointer is replaced with input tensor pointer : https://github.com/openvinotoolkit/openvino/blob/648351ba391ce55e17bcc6e758d824a9220f4711/src/core/src/op/result.cpp#L34
So if some tensor name was set in Result output tensor, it becomes lost. 
If we try to set tensor name again after `validate_nodes_and_infer_types()`, it works, as pointers are the same at this point.

The issue appeared when `cf=True` was turned on in MOC transformations.
`cf=True` turns on shape-off constant folding and removes `validate_nodes_and_infer_types()` in the end of MOC transformations pipeline: https://github.com/openvinotoolkit/openvino/blob/648351ba391ce55e17bcc6e758d824a9220f4711/src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp#L258

Solution: 
Add `validate_and_infer_types()` in the end of MOC pipeline if `cf=True`.

Ticket: CVS-117068

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names
* [x]  Transformation preserves tensor names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests
* [x]  Transformation tests
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list
* [x]  Guide on how to convert the **public** model
* [x]  User guide update

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp [hunk 1]
```diff
@@ -256,6 +256,10 @@ bool ov::pass::MOCTransformations::run_on_model(const std::shared_ptr<ngraph::Fu
             param->set_partial_shape(input_shapes.at(param.get()));
         }
         f->validate_nodes_and_infer_types();
+    } else {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp:259]: could you add a comment explaining why do we need it here?

**Comment by itikhono** [comment]
Not so sure that this is a correct fix. 
I believe Validation pass should work after Constant folding if CF actually happens and returns the value "true" from the function

File: src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp [hunk 2]
```diff
@@ -256,6 +256,10 @@ bool ov::pass::MOCTransformations::run_on_model(const std::shared_ptr<ngraph::Fu
...
+        for (auto&& result : f->get_results()) {
```
[the remark this answers is not in the document]
Comment (popovaan) [reply src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp]: The solution was changed to run extra validate_nodes_and_infer_types().

File: src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp (hunk 1 above)
Comment (popovaan) [reply src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp]: Removed it.

**Comment by popovaan** [comment]
> Not so sure that this is a correct fix. I believe Validation pass should work after Constant folding if CF actually happens and returns the value "true" from the function

As discussed added validate_nodes_and_infer_types() at the end of MOC pipeline.

File: src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp [hunk 3]
```diff
@@ -255,8 +255,8 @@ bool ov::pass::MOCTransformations::run_on_model(const std::shared_ptr<ngraph::Fu
         for (auto&& param : f->get_parameters()) {
             param->set_partial_shape(input_shapes.at(param.get()));
         }
-        f->validate_nodes_and_infer_types();
     }
+    f->validate_nodes_and_infer_types();
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp:259]: could you add c++ test?
with Add->Add->Add + LinOp fusion transformation and check that we can set the names

Comment (popovaan) [reply src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp:259]: Done.

File: src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (popovaan) [reply src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp]: Added cpp test as well.

**Review by itikhono — APPROVED** [review]

**Review by rkazants — APPROVED** [review]

**Pull request merged**